### ISSUE TITLE
Does anyone know if there are CPU HEVC Limitations?

### ISSUE THREAD

**Issue by TheSashmo** [CESNET/UltraGrid#206]
I've been using UG for a while now, and resorted to always using CPU H.264 encoding/decoding.

I have a 48 core Xeon (Intel(R) Xeon(R) CPU E5-2678 v3 @ 2.50GHz) machine that should be able to comfortably encode at least two HEVC on CPU without having any frame dropped due to CPU limitations.

I can see that when I start my first session, it works well and I see the frame count sold at 150 frames every 5 seconds.

If I start another thread, I see momentary drop in frames on the first session, (obviously as it hits the first startup load), but after a bit I can see it fluctuate and both sessions don't consistently stay at the desired 150 frames every 5 seconds.  I've set the lowest sub sampling i.e. 4:2:0.  Assuming that the machine even though 48 cores hyper threaded, really  means 24 cores, I can see that the load never goes above 24 when trying to encode two signals, but I can still see the occasional frame drop below 150.

I know that 149/150/151 is acceptable, BUT, I can see at times it drops even further into the 140's, and I can measure that those frames are missing on the output decode, so I know for sure that input vs. output, are not 100% matched

Testing using decklink and 1080i59.94 UYVY

Am I missing something?

`-c libavcodec:encoder=libx265:bitrate=4000k:subsampling=420 -s embedded --audio-codec=AAC:sample_rate=48000:bitrate=128k --audio-capture-format channels=2 127.0.0.1 -l unlimited -m 1316 -P 11000`

I have tested, and would consider using hardware acceleration via nvenc, but the additional latency doesn't help my solution.  i.e. latency vs bandwidth saved using H265 vs H264.

**Comment by MartinPulec** [comment]
Hi, I assume that the server is a NUMA? If so, I'd suggest to set CPU affinity for the different UG processes. What operating system it is? It should be similar also for different systems, but for Linux, could you try something like:

    taskset --cpu-list 0-11  <uv1>
    taskset --cpu-list 12-23 <uv2>

This should definitely improve stability of UltraGrid. It is uncertain if it improves performance since it uses only ¼ of cores. Maybe you could try ranges `0-23` and `24-47` then. But I would start with use of singe CPU instead because of the characteristics of NUMA - using cores of single processor mean that only RAM local to that CPU is used.

CPU affinity [can be set also in Windows](https://github.com/CESNET/UltraGrid/wiki/UltraGrid-FAQ#how-to-optimize-performance-for-a-numa).

**Comment by TheSashmo** [comment]
Yeah I have fine memories of doing this years ago.  Unfortunately that doesn't solve the problem.  I will experiment more, maybe disabling hyper-threading, and trying again.

**Comment by MartinPulec** [comment]
That's a pity, this seemed to be a good suspect to me. I don't thing there should be any dependency between those two UltraGrid instances that should influence this.

Just another related idea (that more likely won't help but can be tried) - what is the input signal it is some PCIe card? Apart from affinity, individual PCIe cards are wired to certain CPU. So it might help to 1) plug each card to a separate CPU and let each UG instance to run on that specified CPU. Do you think that it might help? I don't recall how to check to which CPU a particular card is connected but I can look.

Regarding NVENC mentioned in first post - it perhaps adds some latency but I wouldn't say much. But there can be additional glitches that need to be checked.

**Comment by TheSashmo** [comment]
I think is worth checking into, but I honestly just think now that the HEVC compression during high motion needs MORE power.  If I have a low moving image, it's almost flawless in the frame rate processed messages.  If I change it up to a high motion test video, you can see the frames are dropped almost right away.   Are my assumptions correct when I talk about 150 frames in 5 seconds?  Meaning is that the correct or ideal way to measure that I am doing things correctly on the encoder side?

**Comment by MartinPulec** [comment]
Sounds legit. And yes, I'd assume that in this scenario here the frame rate really measures the performance of the compression. You could also use `--verbose=debug`, which prints you individual frames' compression duration.

**Comment by MartinPulec** [comment]
Closing for inactivity, feel free to reopen with eventual updates.
